dbc2val/overlay/README.md

```diff
@@ -0,0 +1,219 @@
+# Using overlays as alternative mapping
+
+## Example 1 - Boolean
+
+```
+VCFRONT_passengerPresent:
+  vss: # taken from https://github.com/COVESA/vehicle_signal_specification/blob/master/spec/Cabin/SingleSeat.vspec
+    datatype: boolean 
+    type: sensor
+    unit: none
+    description: Does the seat have a passenger in it.
+  databroker:
+    datatype: 1 # BOOL number taken from types_pb2.py
+    changetype: 2 # ON_CHANGE number taken from types_pb2.py
+  minupdatedelay: 1000
+  targets:
+    Vehicle.Cabin.Seat.Row1.Pos2.IsOccupied:
+      transform:
+        fullmapping:
+          1: "true"
+          0: "false"
+```
+
+A functionally equal mapping could possibly be like this:
+
+```
+Vehicle.Cabin.Seat.Row1.Pos2.IsOccupied:
+  type: actuator
+  datatype: boolean
+  dbc: VCFRONT_passengerPresent
+  dbc_mapping: [{1:true},{0:false}]
```

Comment:
Well we could actually have something very similar to how it was before, like


```
  transform:
    fullmapping:
      "1": true
      "0": false
```

(Current VSS-tools cannot handle integers as keys, but if we include them in quotation marks it works well)

So in short - we can represent mapping as a tree structure - did not thought that was possible

Reply:
I added a separate review comment, but maybe a reply here fits better into the discussion..

I don't think it's a good idea to put what is essentially a "value" as a key like this. Keys are (usually) used to describe the corresponding value, which is one reason restrictions like "keys can only be strings" can make sense.

The mapping can be specified as something like this instead:
```json
  "mapping": [
    {
      "from": 0,
      "to": false
    },
    {
      "from": 1,
      "to": true
    }
  ]
```
which would avoid the issue. Since this uses the keys to describe the values, the direction of the mapping is more clear as well (I think).

Reply:
```  
transform:
    fullmapping:
      "1": true
      "0": false
```

If this works (or the equivalent of johns proposal), by "accident" in vss-tools, the question really is, do we want to allow arbitrary structures under extended attributes from VSS point of view? It might become painful for tooling

If it is seen "just as a string", (that my have some internal yaml/json/whatever structure) it might be easier, even if maybe not really "elegant" in the computer science sense...

(Not sure this is better discussed VSS upstream...)

On the other hand _if_ structures (as in "YAML in the originating vspec") are acceptable, probably all should go under one extended attribute, i.e. `dbc`

```yaml
Vehicle.Cabin.Seat.Row1.Pos2.IsOccupied:
  type: sensor
  datatype: boolean
  dbc:
    signal: VCFRONT_passengerPresent
    transform:
      mapping:
        - from: 0
          to: false
        - from: 1
          to: true
```

Reply:
I think it would be really beneficial if everything that is not "reserved" for use in VSS, was just passed along unmodified (to be used by the intended downstream tools or applications).

For YAML and JSON, it seems to work "accidentally" already..

For other formats, it would probably need some type of "extended attributes" and a `Variant` type holding the values of them, but maybe it doesn't have to be supported by all output formats anyway..